Modules/_winapi.c

```diff
@@ -744,6 +744,20 @@ getenvironment(PyObject* environment)
                 "environment can only contain strings");
             goto error;
         }
+        if (PyUnicode_FindChar(key, '\0', 0, PyUnicode_GET_LENGTH(key), 1) != -1 ||
+            PyUnicode_FindChar(value, '\0', 0, PyUnicode_GET_LENGTH(value), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "embedded null character");
+            goto error;
+        }
+        /* Search from index 1 because on Windows starting '=' is allowed for
+           defining hidden environment variables. */
+        if (PyUnicode_GET_LENGTH(key) == 0 ||
+            PyUnicode_FindChar(key, '=', 1, PyUnicode_GET_LENGTH(key), 1) != -1)
```

Comment:
@serhiy-storchaka I am curious about this. Is there a particular version of Windows this works on? We copied this code in PyPy, but it is failing tests for `putenv("=hidden", "foo")`

Reply:
@mattip, the system tracks the working directory on each drive using environment variable names that start with "=", such as "=C:". Python's [`os.chdir()`](https://github.com/python/cpython/blob/v3.9.6/Modules/posixmodule.c#L1725) implements this.  Such names are accessible via WinAPI `GetEnvironmentStringsW`, `GetEnvironmentVariableW`, and `SetEnvironmentVariableW`. In most contexts such variables are hidden or filtered out, and using them at the application level isn't officially supported. In particular, CRT environment functions such as `putenv` and `getenv` do not support them, and they're filtered out of C `[_w]environ`, from which Python initializes `os.environ`.

Here's a manual example:

```python
>>> os.path.abspath('Z:')
'Z:\\'
>>> win32api.SetEnvironmentVariable('=Z:', 'Z:\\test_cwd')
>>> os.path.abspath('Z:')
'Z:\\test_cwd'
```

Or set the drive working directory in the environment of a child process:

```python
>>> os.path.abspath('Z:')
'Z:\\'
>>> environ = os.environ.copy()
>>> environ['=Z:'] = 'Z:\\test_cwd'
>>> subprocess.call('python -c "import os;print(os.path.abspath(\'Z:\'))"', env=environ)
Z:\test_cwd
0
```

Reply:
Thanks for the quick reply. So if I understand correctly while the code (in three places: this one and twice in `posix_module`) may permit the use of `=hidden`, it is not guaranteed that the subsequent system call will succeed.

Reply:
Names that begin with "=" always work with the Windows functions `GetEnvironmentVariableW` and `SetEnvironmentVariableW`, and they're inherited fine by child processes. For example:

```python
>>> win32api.SetEnvironmentVariable('=spam', 'eggs')
>>> win32api.GetEnvironmentVariable('=spam')
'eggs'
>>> subprocess.call('python -c "import win32api; print(win32api.GetEnvironmentVariable(\'=spam\'))"')
eggs
0
```
However, such names are not officially supported. IMO, this is primarily because the C runtime filters them out of its `[_w]environ` global variable and doesn't support them in environment functions such as `putenv()`. This in turn is because C and POSIX are tightly linked, and [POSIX disallows](https://pubs.opengroup.org/onlinepubs/000095399/basedefs/xbd_chap08.html) "=" in environment variable names. Python's `os.environ`, `os.getenv()`, and `os.putenv()` are based on the C runtime, so they have the same limits, and probably should since these are C/POSIX constructs. Maybe one day Python will provide native Windows functionality in the os module that directly accesses the environment block of the process and directly exposes the WinAPI `GetEnvironmentVariableW` and `SetEnvironmentVariableW` functions. For now, that's only possible with extensions such as PyWin32 or ctypes.

Reply:
Thanks. So I will keep PyPy's `putenv` consistent with CPython: python allows `=hidden` as a name, but then the `putenv` syscall will fail.